### ISSUE TITLE
Not so fast and the performance is not as good as the paper declares

### ISSUE THREAD

**Issue by ZhenYangIACAS** [awslabs/sockeye#263]
I find that the training is not so fast and it takes 4 days to train the wmt16en-de,  however, the tensor2tensor only needs 2 days. I do not know why the sockeye needs 1M updates? As I set the batch_size as 4096 both for the sockeye and tensor2tensor, the tensor2tensor runs 6000 updates for one epoch, but the sockeys runs  30000 updates for one epoch. The script for running the sockeye is：
 export CUDA_VISIBLE_DEVICES='8,9,10,11,12,13'

python3 -m sockeye.train \
-s /data/zhyang/dl4mt/corpus/data_450w_en_de/transformer/450w_ende_data/train.tok.clean.bpe.32000.en \
-t /data/zhyang/dl4mt/corpus/data_450w_en_de/transformer/450w_ende_data/train.tok.clean.bpe.32000.de \
-vs /data/zhyang/dl4mt/corpus/data_450w_en_de/transformer/450w_ende_data/newstest2013.tok.bpe.32000.en \
-vt /data/zhyang/dl4mt/corpus/data_450w_en_de/transformer/450w_ende_data/newstest2013.tok.bpe.32000.de \
 -o en-de \
 --seed=1 --batch-type=word --batch-size=4096 --checkpoint-frequency=4000 --device-ids=-6 --embed-dropout=0:0 --encoder=transformer --decoder=transformer --num-layers=6:6 --transformer-model-size=512 --transformer-attention-heads=8 --transformer-feed-forward-num-hidden=2048 --transformer-preprocess=n --transformer-postprocess=dr --transformer-dropout-attention=0.1 --transformer-dropout-act=0.1 --transformer-dropout-prepost=0.1 --transformer-positional-embedding-type fixed --fill-up=replicate --max-seq-len=100:100 --label-smoothing 0.1 --weight-tying --weight-tying-type=src_trg_softmax --num-embed 512:512 --num-words 50000:50000 --word-min-count 1:1 --optimizer=adam --optimized-metric=perplexity --initial-learning-rate=0.0001 --learning-rate-reduce-num-not-improved=8 --learning-rate-reduce-factor=0.7 --learning-rate-scheduler-type=plateau-reduce --learning-rate-warmup=0 --max-num-checkpoint-not-improved=32 --min-num-epochs=0 --max-updates 1001000 \
--weight-init xavier --weight-init-scale 3.0 --weight-init-xavier-factor-type avg

After finished training, I only get 26.01 BLEU on newstest2014. However, the tensor2tensor gets 26.75.

**Comment by tdomhan** [comment]
Looking at the parameters you were using for training I can spot two differences to the ones used in the experiments for the paper (see https://github.com/awslabs/sockeye/blob/arxiv_1217/arxiv/code/transformer/sockeye/train-transformer.sh). Namely:
 * `--clip-gradient=-1`
 * `--transformer-dropout-relu`

Now these of course won't have any influence on the number of batches, but they might affect your final BLEU score.
How did you run BLEU evaluation? Did you run both tensor2tensor and Sockeye through [ScareBLEU](https://github.com/awslabs/sockeye/blob/master/contrib/sacrebleu/sacrebleu.py)?

With regards to the updates per epoch: the training data has around 160 million words. With a batch size of 4096 this gives around 39,000 batches. If you only see 6000 updates per epoch in tensor2tensor you are probably using a batch size larger than 4096 words per batch. With the 160 million source words the batch size most probably be something like 26k words. Could you share your tensor2tensor configuration?

Also could you paste your sockeye training log for reference?

**Comment by tdomhan** [comment]
I will close the issue for now due to inactivity. For now it seems like the two systems were configured differently. Feel free to reopen.
